KAFKA-14915: allow read remote storage for multiple partitions in one fetchRequest

showuon opened the pull request:
This PR enables reading remote storage for multiple partitions in one
fetchRequest. The main changes are:
1. In `DelayedRemoteFetch`, we accept multiple remoteFetchTasks and
other metadata now.
2. In `DelayedRemoteFetch`, we'll wait until all remoteFetch done,
either succeeded or failed.
3. In `ReplicaManager#fetchMessage`, we'll create one
`DelayedRemoteFetch` and pass multiple remoteFetch metadata to it, and
watch all of them.
4. Add tests

showuon committed:
KAFKA-14915: allow read multiple partitions in one fetchRequest

showuon committed:
KAFKA-14915: fix broken tests

showuon committed:
KAFKA-14915: add tests

showuon committed:
KAFKA-14915: update javadoc

showuon committed:
KAFKA-14915: add more tests

showuon commented:
@satishd @kamalcph , please take a look. Thanks.

showuon committed:
KAFKA-14915: improve tests

kamalcph commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -81,17 +82,23 @@ class DelayedRemoteFetch(remoteFetchTask: Future[Void],
-      forceComplete()
-    else
+    // Case c
+    if (remoteFetchResults.values().stream().anyMatch(taskResult => !taskResult.isDone))
```
nit: subjective
Can the condition be inverted?

```
// Case c
if (remoteFetchResults.values().stream().allMatch(taskResult => taskResult.isDone))
  forceComplete()
else
  false
```

kamalcph commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1598,31 +1601,40 @@ class ReplicaManager(val config: KafkaConfig,
+        // Store the error in RemoteLogReadResult if any in scheduling the remote fetch task.
+        // It will be sent back to the client in DelayedRemoteFetch along with other successful remote fetch results.
+        remoteFetchResult.complete(new RemoteLogReadResult(Optional.empty, Optional.of(e)))
+        delayedRemoteFetchPurgatory.checkAndComplete(key)
```
DelayedRemoteFetch is not instantiated at this place. Calling the purgatory#checkAndComplete might complete the existing delayed operation in the purgatory (which is fine). Could you explain why do we need this change?
```
delayedRemoteFetchPurgatory.checkAndComplete(key)
```

showuon replied:
You're right! This should be removed. Thanks for catching it.

kamalcph commented on core/src/test/scala/integration/kafka/server/DelayedRemoteFetchTest.scala:
```diff
@@ -170,52 +206,252 @@ class DelayedRemoteFetchTest {
+    assertTrue(responses.contains(topicIdPartition2))
+
+    assertEquals(Errors.NONE, responses(topicIdPartition).error)
+    assertEquals(Errors.NONE, responses(topicIdPartition2).error)
```
can the highWatermark and logStartOffset also be asserted? It should also be filled in the response.

kamalcph commented on core/src/test/scala/integration/kafka/server/DelayedRemoteFetchTest.scala:
```diff
@@ -235,7 +471,8 @@ class DelayedRemoteFetchTest {
     new LogReadResult(
-      new FetchDataInfo(LogOffsetMetadata.UNKNOWN_OFFSET_METADATA, MemoryRecords.EMPTY),
+      new FetchDataInfo(LogOffsetMetadata.UNKNOWN_OFFSET_METADATA, MemoryRecords.EMPTY, false, Optional.empty(),
+        Optional.of(mock(classOf[RemoteStorageFetchInfo]))),
```
Thanks for making this change! Previously, it was reading the results from `localReadResults`.

showuon replied:
You're right! We did not test what we expected to be tested. :)

kamalcph reviewed (commented):
Thanks for the patch! Overall LGTM. 

Few points: 

1. The FETCH request now has to wait to get response from all the remote partitions. And, it can have tail-latency. This tradeoff is fine considering the consumer might have to read the same data in the upcoming FETCH requests. This can be documented in the PR. 
2. When `minOneMessage` set to false (from second partition in the request), and there are multiple remote reads in the given FETCH request. We may read the data from remote and then drop those when the size of the message does not fit the partition limit. This was fine for local-reads but remote-reads might fetch chunk of data and multiple responses might be dropped. Shall we add some DEBUG logs when the remote fetched data gets dropped?

showuon committed:
KAFKA-14915: Improve docs and address comments

showuon commented:
> 1. The FETCH request now has to wait to get response from all the remote partitions. And, it can have tail-latency. This tradeoff is fine considering the consumer might have to read the same data in the upcoming FETCH requests. This can be documented in the PR.

Good point. I've updated the doc in `remote.fetch.max.wait` to mention the waiting all fetch behavior, and reverted the change in this [PR](https://github.com/apache/kafka/pull/19336) about the limitation.
> 2. When minOneMessage set to false (from second partition in the request), and there are multiple remote reads in the given FETCH request. We may read the data from remote and then drop those when the size of the message does not fit the partition limit. This was fine for local-reads but remote-reads might fetch chunk of data and multiple responses might be dropped. Shall we add some DEBUG logs when the remote fetched data gets dropped?

I agree with you. Added a debug log.

Thanks for the review!

satishd commented:
Thanks @showuon for the PR, I will review this PR by Monday.

kamalcph commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -90,8 +92,13 @@ class DelayedRemoteFetch(remoteFetchTask: Future[Void],
-    val cancelled = remoteFetchTask.cancel(false)
-    if (!cancelled) debug(s"Remote fetch task for RemoteStorageFetchInfo: $remoteFetchInfo could not be cancelled and its isDone value is ${remoteFetchTask.isDone}")
+    remoteFetchTasks.forEach { (topicIdPartition, task) =>
+      if (!task.isDone) {
```
should we add a null check? The task can be null when thread-pool throws RejectedExecutionException in ReplicaManager#processRemoteFetch.

```
if (task != null && !task.isDone)
```

fvaleri commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1901,7 +1905,7 @@ class ReplicaManager(val config: KafkaConfig,
           // For consume fetch requests, create a dummy FetchDataInfo with the remote storage fetch information.
           // For the first topic-partition that needs remote data, we will use this information to read the data in another thread.
```
Should we also update this comment?

showuon committed:
KAFKA-14915: address review comments

kamalcph reviewed (approved):
LGTM! Thanks for the patch!

fvaleri reviewed (approved):
LGTM. Thanks.

I tried on my local machine and I can see remote partitions being read in parallel and increased throughput.

satishd commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3586,6 +3586,109 @@ class ReplicaManagerTest {
+      assertEquals(2, capturedFetchInfos.size, "Should have 2 remote storage fetch info calls")
+
+      val capturedTopicPartitions = capturedFetchInfos.map(_.topicIdPartition.topicPartition).toSet
+      assertTrue(capturedTopicPartitions.contains(tp0), "Should contain tp0")
```
assert condition should have topic partition: "Should contain tp0" -> "Should contain " + tp0

satishd commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3586,6 +3586,109 @@ class ReplicaManagerTest {
+
+      val capturedTopicPartitions = capturedFetchInfos.map(_.topicIdPartition.topicPartition).toSet
+      assertTrue(capturedTopicPartitions.contains(tp0), "Should contain tp0")
+      assertTrue(capturedTopicPartitions.contains(tp1), "Should contain tp1")
```
assert condition should have topic partition: "Should contain tp1" -> "Should contain " + tp1

satishd commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3586,6 +3586,109 @@ class ReplicaManagerTest {
+
+      // Verify that response contains both tidp0 and tidp1 and have no errors
+      val responseTopicIdPartitions = responseData.map(_._1).toSet
+      assertTrue(responseTopicIdPartitions.contains(tidp0), "Response should contain tidp0")
```
assert condition should have topic partition: "Response should contain tidp0" -> "Response should contain" + tidp0

satishd commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3586,6 +3586,109 @@ class ReplicaManagerTest {
+      // Verify that response contains both tidp0 and tidp1 and have no errors
+      val responseTopicIdPartitions = responseData.map(_._1).toSet
+      assertTrue(responseTopicIdPartitions.contains(tidp0), "Response should contain tidp0")
+      assertTrue(responseTopicIdPartitions.contains(tidp1), "Response should contain tidp1")
```
assert condition should have topic partition: "Response should contain tidp1" -> "Response should contain" + tidp1

satishd reviewed (commented):
Thanks @showuon for the PR, overall LGTM.

Left a couple of comments on minor typos int he test.

showuon committed:
KAFKA-14915: improve the test error output

satishd reviewed (approved):
Thanks @showuon for addressing the comments. LGTM.
